x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/stringstats/InternalStringStats.java

```diff
@@ -0,0 +1,280 @@
+        out.writeNamedWriteable(format);
+        out.writeBoolean(showDistribution);
+        out.writeVLong(count);
+        out.writeVLong(totalLength);
+        out.writeVInt(minLength);
+        out.writeVInt(maxLength);
+        out.writeMap(charOccurrences, StreamOutput::writeString, StreamOutput::writeLong);
+    }
+
+    public String getWriteableName() {
+        return StatsAggregationBuilder.NAME;
+    }
+
+    public long getCount() {
+        return count;
+    }
+
+    public int getMinLength() {
+        return minLength;
+    }
+
+    public int getMaxLength() {
+        return maxLength;
+    }
+
+    public double getAvgLength() {
+        return (double) totalLength / count;
+    }
+
+    public double getEntropy() {
+        double sum = 0.0;
+        double compensation = 0.0;
+        for (double p : getDistribution().values()) {
+            if (p > 0) {
+                // Compute the sum of double values with Kahan summation algorithm which is more
+                // accurate than naive summation.
+                double value = p * log2(p);
+                if (Double.isFinite(value) == false) {
+                    sum += value;
+                } else if (Double.isFinite(sum)) {
+                    double corrected = value - compensation;
+                    double newSum = sum + corrected;
+                    compensation = (newSum - sum) - corrected;
+                    sum = newSum;
+                }
+            }
+        }
+        return -sum;
+    }
+
+    /**
+     * Convert the character occurrences map to character frequencies.
+     *
+     * @return A map with the character as key and the probability of
+     * this character to occur as value. The map is ordered by frequency descending.
+     */
+    public Map<String, Double> getDistribution() {
+       return charOccurrences.entrySet().stream()
+            .sorted((e1, e2) -> e2.getValue().compareTo(e1.getValue()))
+            .collect(
+                Collectors.toMap(e -> e.getKey(), e -> (double) e.getValue() / totalLength,
+                    (e1, e2) -> e2, LinkedHashMap::new)
+            );
+    }
+
+    /** Calculate base 2 logarithm */
+    static double log2(double d) {
+        return Math.log(d) / Math.log(2.0);
+    }
+
+    public String getCountAsString() {
+        return format.format(getCount()).toString();
+    }
+
+    public String getMinLengthAsString() {
+        return format.format(getMinLength()).toString();
+    }
+
+    public String getMaxLengthAsString() {
+        return format.format(getMaxLength()).toString();
+    }
+
+    public String getAvgLengthAsString() {
+        return format.format(getAvgLength()).toString();
+    }
+
+    public String getEntropyAsString() {
+        return format.format(getEntropy()).toString();
+    }
+
+    public Object value(String name) {
+        Metrics metrics = Metrics.valueOf(name);
+        switch (metrics) {
```

Comment:
Bit of a nit, but I don't love switching on an enum.  If someone later adds a field to the enum, they need to remember to also update this switch.  IMHO, a better solution would be to put a method on the enum `getFieldValue(InternalStringStats stats)` which could then call the appropriate getter and return the value.  That way, any new enum value would need to implement the method for it to compile.